# Week 02 Assignment glucose level data

Welcome to week two of this course programming 1. You will learn about time related data wrangling with pandas and you will learn to visualize with bokeh. This week will be focussed around missing data. Concretely, you will preprocess the glucose json file using interpolation to impute in order to conduct visual analysis. Learning outcomes:

- load a json dataset 
- typecast the Pandas DataFrame to appropiate data types
- inspect the dataset for quality and metadata information
- add a column with interpolated data in Pandas DataFrame
- perform visual analysis

The assignment consists of 6 parts:

- [part 1: load the data](#0)
     - [Exercise 1.1](#ex-11)
- [part 2: prepare for inspection](#1)
     - [Exercise 2.1](#ex-21)
- [part 3: inspect the data](#2)
     - [Exercise 3.1](#ex-31)
- [part 4: interpolate the data](#3)
     - [Exercise 4.1](#ex-41)
- [part 5: visualize the data](#4)
     - [Exercise 5.1](#ex-51)
- [part 6: Challenge](#5)
     - [Exercise 6.1](#ex-61)

Part 1 and 5 are mandatory, part 6 is optional (bonus)
To pass the assingnment you need to a score of 60%. 


<a name='0'></a>
## Part 1: Load the data

Instructions: Load the json datafile `glucose.json` into a pandas dataframe. Check your dataframe with a `.head()` to compare with the expected outcome

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>json.load() method reads a file, pd.read_json converts it to a Pandas DataFrame</li>
    <li>when loading into a Pandas DataFrame use records orientation </li>
</ul>
</details>

<a name='ex-11'></a>
### Code your solution

In [246]:
# CODE YOUR SOLUTION HERE
import numpy as np
import pandas as pd
import json


with open('data/glucose.json') as f:
    f = json.load(f)    #f is a string
    data = pd.read_json(f)
    # data = json.loads(f)    #data is a list
    # data = json.loads(json.load(f))
# print(data)
# print(type(data))

df = pd.DataFrame(data)
df.head()

# print(data[0].keys())
# a = data['ID'][0]
# print(a.keys())


# f
# data = json.load(f)
# data
# print(type(data))
# print(data.keys())

ID              time  recordtype glucose
0  2845.0  2019-04-25 00:08           1     109
1  2850.0  2019-04-25 00:50           1        
2  2877.0  2019-04-25 07:02           1     123
3  2881.0  2019-04-25 07:34           1     158
4  2886.0  2019-04-25 08:19           1

#### Expected outcome: 

<a name='1'></a>
## Part 2: Prepare the data

Check the datatypes of your dataframe. The `glucose` field should be an integer, the `time` field should have a datetime format. If the datatypes are different you should typecast them to the right format.
Make sure that your dataset is sorted by the time column


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use astype() method or pandas.DataFrame.to_datetime() for instance</li>
    <li>make sure that the empty spaces are filled with NaN. Use errors='coerce'</li>
    <li>set_index(), sort_index() and reset_index() are helpful to sort on index</li>
</ul>
</details>

<a name='ex-21'></a>
### Code your solution

In [247]:
# CODE YOUR SOLUTION HERE

## replace empty string with np.nan so it can be converted to float
df = df.replace(r'^\s*$', np.nan, regex=True)

## change ID to int, time to datetime64[ns] and glucose to float
df = df.astype({"ID": 'int64', "time": 'datetime64[ns]', 'glucose': float})
df = df.sort_values('time',axis = 0, ignore_index=True)
columns_titles = ['time', 'ID', 'recordtype', 'glucose']
df=df.reindex(columns=columns_titles)
print(df.dtypes)
# df = df.set_index(['time']).sort_index()
print(df)


time          datetime64[ns]
ID                     int64
recordtype             int64
glucose              float64
dtype: object
                   time                   ID  recordtype  glucose
0   2019-04-25 00:08:00                 2845           1    109.0
1   2019-04-25 00:14:00 -9223372036854775808           0      NaN
2   2019-04-25 00:29:00 -9223372036854775808           0      NaN
3   2019-04-25 00:44:00 -9223372036854775808           0      NaN
4   2019-04-25 00:50:00                 2850           1      NaN
..                  ...                  ...         ...      ...
131 2019-04-25 23:02:00 -9223372036854775808           0      NaN
132 2019-04-25 23:18:00 -9223372036854775808           0      NaN
133 2019-04-25 23:31:00                 3062           1    111.0
134 2019-04-25 23:33:00 -9223372036854775808           0      NaN
135 2019-04-25 23:48:00 -9223372036854775808           0      NaN

[136 rows x 4 columns]


#### Expected outcome: 

<a name='2'></a>
## Part 3: Inspect the data

Now that we prepared the data we are going to inspect the data to get more familiar with the data. You are required to do the following

- inspect the percentage missing data for glucose
- what is the relationship between recordtype and glucose value?
- what is the relationship between ID and glucose value?

Code the solutions to your answers. Create meaningful overviews or statistics

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>In the week 01 assignment some functions were explained to inspect missing values</li>
    <li>In the week 01 assignment some functions were explained to groupby value</li>
</ul>
</details>

<a name='ex-31'></a>
### Code your solution

In [248]:
#CODE YOUR SOLUTION HERE
# print(df)

percent_missing = df['glucose'].isnull().sum() / len(df)
print(percent_missing)




0.6176470588235294


#### Expected outcome percentage missing data
0.6176470588235294

<a name='3'></a>
## Part 4: Interpolate the data

A lot of data is missing. Use interpolation to fill the missing values. Create a new column with the interpolated data. Take an argumentative approach. Select an interpolation method that suits the nature of the data and explain your choice. Mind you that the expected outcome of the interpolation values can differ from the example below

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use Pandas.DataFrame.interpolate() method</li>
</ul>
</details>

<a name='ex-41'></a>
### Code your solution

In [249]:
#CODE YOUR SOLUTION HERE

interpmethod = ['nearest', 'zero', 'slinear', 'quadratic', 'cubic', 'spline', 'barycentric', 'polynomial']

df['interpolated'] = df['glucose'].interpolate()
# df['interpolated2'] = df['glucose'].interpolate(method = 'nearest')
# df['interpolated3'] = df['glucose'].interpolate('index')
# df['interpolated4'] = df['glucose'].interpolate('krogh')


print(df.head())


                 time                   ID  recordtype  glucose  interpolated
0 2019-04-25 00:08:00                 2845           1    109.0    109.000000
1 2019-04-25 00:14:00 -9223372036854775808           0      NaN    109.466667
2 2019-04-25 00:29:00 -9223372036854775808           0      NaN    109.933333
3 2019-04-25 00:44:00 -9223372036854775808           0      NaN    110.400000
4 2019-04-25 00:50:00                 2850           1      NaN    110.866667


#### Example outcome

<a name='4'></a>
## Part 5: Plot the data

Create a plot with the original data and the interpolated data. Consider what the best representation is for visualisation of actual values and modelled/imputed values. An example of such a plot is given below. This plot however is not considered the best practice. 

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>figure(x_axis_type='datetime') automatically makes nices labels of the datetime data</li>
</ul>
</details>

<a name='ex-51'></a>
### Code your solution

In [250]:
from bokeh.io import output_notebook
from bokeh.plotting import figure, show
from bokeh.layouts import gridplot
from bokeh.io import output_notebook
from bokeh.plotting import ColumnDataSource
from bokeh.models import CustomJS, Dropdown
output_notebook()

Loading BokehJS ...

In [251]:
#CODE YOUR SOLUTION HERE

## Rolling mean
df['rolling mean'] = df['interpolated'].rolling(7).mean()


source = ColumnDataSource(df)
p = figure(x_axis_type = 'datetime')
p.line(x='time', y='interpolated', line_width = 2, source=source, color = 'red', legend_label='Interpolated')
# p.line(x='time', y='interpolated2', line_width = 2, source=source, color = 'red', legend_label='Interpolated2')
# p.line(x='time', y='interpolated3', line_width = 2, source=source, color = 'red', legend_label='Interpolated3')
# p.line(x='time', y='interpolated4', line_width = 2, source=source, color = 'red', legend_label='Interpolated4')
p.triangle(x='time', y='glucose', line_width = 4, source=source, color = 'blue', legend_label='Data')
p.line(x='time', y='rolling mean', line_width = 2, source=source, color = 'green', legend_label='Rolling mean')
p.title.text = 'Click on legend entries to hide the corresponding lines'

p.legend.location = "top_left"
p.legend.click_policy="hide"

p.xaxis.axis_label = 'time'
p.yaxis.axis_label = 'glucose'

show(p)

## Add dropdown menu for interpolation options

## Dropdown for interpolation options:
# menu = [("Item 1", "item_1"), ("Item 2", "item_2")]
# dropdown = Dropdown(label="Interpolation type", button_type="warning", menu=menu)
# dropdown.js_on_event("menu_item_click", CustomJS(code="console.log('dropdown: ' + this.item, this.toString())"))

# show(dropdown)






<a name='6'></a>
## Part 6: Challenge

It might even be interesting to introduce a widget in which you can select different methods to interpolate.
1. Can you improve the interpolation by choosing an other method?
2. Can you add an rolling mean line? 
2. Can you improve the plot by making it interactive?

<a name='ex-61'></a>
### Code your solution

In [252]:
#CODE YOUR SOLUTION HERE
